### PR TITLE
Pauleconomy

### PR DESCRIPTION
**Pull request by hetzge**
@paulwedeck Here is the second try. Hope everything is fine now. 

It looks like with this more up to date version the font does not work correctly. There is no % character and the braces look ugly.

### CODE REVIEW

**Comment by paulwedeck** [comment]
This is a known bug.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java [hunk 1]
```diff
@@ -1,66 +1,245 @@
...
-		return ESecondaryTabType.SETTLERS;
+	public void contentShowing(ActionFireable actionFireable) {
+		super.contentShowing(actionFireable);
+		this.uiContentUpdater.start();
 	}
 
 	@Override
-	public void update(IPartitionSettings data) {
+	public void update(IPartitionSettings partitionSettings) {
+		// Use delay to prevent race condition between direct ui update and settings update
```
Comment (paulwedeck) [inline jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:67]: This is an ugly workaround.
Please  refactor SetMoveableRatioAction to not transfer the new value but just the delta.

File: jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java [hunk 2]
```diff
@@ -397,6 +399,11 @@ public void action(IAction action) {
 			case UNLOAD_FERRIES:
 				unloadFerries();
 				break;
+				
+			case SET_MOVEABLE_RATIO:
+				SetMoveableRatioAction ratioAction = (SetMoveableRatioAction) action;
```
Comment (paulwedeck) [inline jsettlers.logic/src/main/java/jsettlers/input/GuiInterface.java:404]: This is where the actual ratio change should take place.
Look at SPEED_FASTER/SPEED_SLOWER for reference

File: jsettlers.logic/src/main/java/jsettlers/input/GuiTaskExecutor.java [hunk 3]
```diff
@@ -227,6 +228,10 @@ public void executeTask(TaskPacket iTask) {
 			case UNLOAD_FERRY:
 				unloadFerry((MovableGuiTask) guiTask);
 				break;
+				
+			case SET_MOVEABLE_RATIO:
+				SetMoveableRatioTask task = (SetMoveableRatioTask) guiTask;
```
Comment (paulwedeck) [inline jsettlers.logic/src/main/java/jsettlers/input/GuiTaskExecutor.java:233]: You could either do something like grid.getMovableRatio()+task.getChange() or grid.changeMovableRatio().
I am not sure what would be the best solution.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java [hunk 4]
```diff
@@ -1,66 +1,245 @@
...
+
+			public void setRatio(float ratio) {
+				this.ratio = ratio;
+			}
+
+			public void setCurrentRatio(float currentRatio) {
+				this.currentRatio = currentRatio;
+			}
+
+			public class UpDownArrows extends Panel {
```
Comment (paulwedeck) [inline jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:173]: This already exists as jsettlers.graphics.ui.CountArrows and does not have the latency issue.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java [hunk 5]
```diff
@@ -1,66 +1,245 @@
...
-	private UIPanel panel;
-	private final UiLocationDependingContentUpdater<IPartitionSettings> uiContentUpdater = new UiLocationDependingContentUpdater<>(this::currentDistributionSettingsProvider);
+	private final ContentPanel panel;
+	private final UiLocationDependingContentUpdater<IPartitionSettings> uiContentUpdater;
+	private long lastChangeTimestamp;
 
 	public ProfessionPanel() {
-		panel = new ProfessionsLayout()._root;
-		uiContentUpdater.addListener(this);
+		this.panel = new ContentPanel();
```
Comment (paulwedeck) [inline jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:36]: All the other Panel are defined in .xml files that are "compiled" to java classes by a preprocessor.
I didn't make this system but I would prefer to keep except if there is a good reason why this panel shouldn't be generated by it of course.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java [hunk 6]
```diff
@@ -1,66 +1,245 @@
...
+		private final SettlerPanel builderPanel;
+
+		public ContentPanel() {
+			super(118f, 216f);
+			this.position = new ShortPoint2D(0, 0);
+			this.carrierPanel = new SettlerPanel(widthInPx, EProfessionType.CARRIER);
+			this.diggerPanel = new SettlerPanel(widthInPx, EProfessionType.DIGGER);
+			this.builderPanel = new SettlerPanel(widthInPx, EProfessionType.BUILDER);
+
+			add(Panel.box(new Label(Labels.getString("Settlers"), EFontSize.NORMAL), widthInPx, 20f), 0f, 0f);
```
Comment (paulwedeck) [inline jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:111]: The Labels.getString is meant as a translation mechanism.
But there is not key "Settlers" defined (atleast not in labels_en.properties or labels_de.properties)
btw. the core game and the mapcreator use different labels_ files for some reason.

This label should probably be translated maybe place it below settler_stats_worker.
There is no general name convention for the labels_ files but I would prefer words as lowercase, _ instead of - and enums as uppercase.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 1 above)
Comment (hetzge) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:67]: I am not sure if this would fix the problem. There is a delay because the change of the ratio is scheduled as a task and the updater of the ui is also working by a timer. I am updating the ui directly so that the user has a fast feedback for his click and after the click this workaround prevents from getting an old update to early (as example if the user clicks multiple times very fast).

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java [hunk 7]
```diff
@@ -1,66 +1,245 @@
...
-	public enum ConfigurableRatioType {
-		BEARER,
-		DIGGER,
-		BRICKLAYER
+	public enum EProfessionType {
+		CARRIER("Carriers", true, EMovableType.BEARER),
+		DIGGER("Diggers", false, EMovableType.DIGGER),
+		BUILDER("Builders", false, EMovableType.BRICKLAYER);
+
+		public final String label;
```
Comment (paulwedeck) [inline jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:87]: This is only meant as text in the ui, isn't it?
You can just call Labels#getLabel(EMovableType) on EProfessionType#moveableType to get translated labels

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 1 above)
Comment (paulwedeck) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:67]: Are you sure that the ui values are updated on a timer and not on every frame?
I think the issue behind the delay is that there is some latency between tasks being scheduled and actually executed (for multiplayer purposes). You can see this when you give move orders or change the stock settings

Comment (paulwedeck) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:67]: or even more noticeably in the production settings if you adjust the Barfill for different tools

Comment (hetzge) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:67]: The ui is updated by UiContentUpdater (and there is a delay of 1000). I am not sure if this is the best available way but using it was very confusing. It is very hard to follow the source of the update calls.

File: jsettlers.logic/src/main/java/jsettlers/input/GuiTaskExecutor.java (hunk 3 above)
Comment (hetzge) [reply jsettlers.logic/src/main/java/jsettlers/input/GuiTaskExecutor.java:233]: I think in this case nothing is wrong with using a absolute value. A set call is much more easy to understand and to debug then a add call.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 1 above)
Comment (paulwedeck) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:67]: I looked into it and the delay in UiContentUpdater seams to make a big difference.
This part of the code is not shared with Android so it wouldn't be an issue to lower the delay to maybe 50-100.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 5 above)
Comment (hetzge) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:36]: I have seen the xml way and was very confused. I am really not a fan of it. The ui in settler has not a very complex layout and the additional effort is very high (there is a additional language (xml), there is a additional build step, there is a not type checked relation between xml and java, it is very static). The benefit is only that it calls some constructors for you.

File: jsettlers.logic/src/main/java/jsettlers/input/GuiTaskExecutor.java (hunk 3 above)
Comment (paulwedeck) [reply jsettlers.logic/src/main/java/jsettlers/input/GuiTaskExecutor.java:233]: But an add call is the natural way of translation the arrow symbol to logic calls.
And there seams to be not debugging issues with SPEED_SLOWER and SPEED_FASTER and they don't need timer stuff in the ui to function as intended.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 1 above)
Comment (hetzge) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:67]: The more i am looking at UiContentUpdater the more i think this thing should be removed completely and instead a update method should be added to AbstractContentProvider. Then the main panel could simply manage the timer and call the update method. The UiContentUpdate seems to provide very much unnecessary complexity.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 5 above)
Comment (paulwedeck) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:36]: Yes, the xml stuff is not ideal and very static but it does the x/y/width/height math for you and it allows easier extension of existing panels.
If we ever come to touch this panel again I might make it a xml file but for now we can leave it as is.

File: jsettlers.logic/src/main/java/jsettlers/input/GuiTaskExecutor.java (hunk 3 above)
Comment (hetzge) [reply jsettlers.logic/src/main/java/jsettlers/input/GuiTaskExecutor.java:233]: The timer stuff should have the same effect with a absolute and a relative value.
With debugging issue i mean not the debugging with a tool. Instead i mean debugging with the brain (understanding and tracking the code).
If i have something like setX(10) then i know the value will be 10. If i have something like increament(10) then i dont know if the result is 10 or 20 or -424152. It could be everything depending on the previous state. Also it would be necessary to have decrement method, too. So the code will be much more robust with the setter.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 1 above)
Comment (paulwedeck) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:67]: There are still two types of UiContentUpdater: player and location based. Having an update method per interested ACP sounds like a lot of copy&paste code.

File: jsettlers.logic/src/main/java/jsettlers/input/GuiTaskExecutor.java (hunk 3 above)
Comment (paulwedeck) [reply jsettlers.logic/src/main/java/jsettlers/input/GuiTaskExecutor.java:233]: I thought more in terms of ChangeMovableRatioTask and INCREASE_MOVABLE_RATIO and DECREASE_MOVABLE_RATIO task types/actions.

The ratios should obviously be limited between 0 and 100.

If you really want use an absolute value you should implement it via an barfill and not as arrows.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java [hunk 8]
```diff
@@ -69,9 +70,13 @@ public void update(IPartitionSettings partitionSettings) {
...
 		}
 	}
+	
+	public void setPlayer(IInGamePlayer player) {
+		uiContentUpdater.setPlayer(player);
+	}
 
 	private static IPartitionSettings currentDistributionSettingsProvider(IGraphicsGrid grid, ShortPoint2D position) {
-		return grid.getPlayerIdAt(position.x, position.y) >= 0 ? grid.getPartitionData(position.x, position.y).getPartitionSettings() : null;
+		return grid.getPlayerAt(position.x, position.y).getPlayerId() >= 0 ? grid.getPartitionData(position.x, position.y).getPartitionSettings() : null;
```
Comment (paulwedeck) [inline jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:79]: grid.getPlayerAt() might return null that is the equivalent if a playerId of -1 or lower.
.getPlayerId() will never return anything below 0 but will just throw a NullPointerException.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 1 above)
Comment (hetzge) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:67]: At the moment we have the following steps in the content providers for each updater that is used:
- Instantiate the updater and pass the provider
- Register the listener
- Start the updater
- Update the position of the updater
- Update the player of the updater
- Handle the updates
- Stop the updater
Most of this steps could be done in the MainPanel once. This would also prevent from leaking resources like i did (forgot to call the stop method). There will be no copy paste code. The abstract content provider could provide two methods (onPositionUpdate, onPlayerUpdate) that could be overriden if needed.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 4 above)
Comment (hetzge) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java]: Ok i reused the existing one.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 7 above)
Comment (hetzge) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:92]: Thanks for the tip. I implemented it with the Labels#getName(EMoveableType) method.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 6 above)
Comment (hetzge) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java]: I have added the label to the properties.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 1 above)
Comment (paulwedeck) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:71]: I think this should be the matter of another pr because this aspect of your pr is working as intended and not really related to bearer limit or similar.
I think AbstractContentProvider is meant as an general purpose Content holder and onPositionUpdate and onPlayerUpdate sounds a bit too specific for that.
I would favor some kind of Subclass e.g. PositionDependentContentProvider that has the onPositionUpdate method.
Whats your opinion on that?

File: jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java [hunk 9]
```diff
@@ -380,9 +351,37 @@ private boolean tryToCreateWorker(WorkerCreationRequest workerCreationRequest) {
...
+			}
+		} 
+		return false;
+	}
+		
+	private void updateProfessionSettings() {
+		ProfessionSettings professionSettings = settings.getProfessionSettings();
+		professionSettings.resetCount();
+	
+		for (ILogicMovable mov : MovableManager.getAllMovables()) {
```
Comment (paulwedeck) [inline jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java:368]: I know that you just reused this code from my initial hack but wouldn't it make more sense to do this loop one time for all partitions instead of one iteration per partition every second?

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 1 above)
Comment (hetzge) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:71]: Yes i totally agree with the separate pull request for this topic.

Additional subclasses like PositionDependentContentProvider is not a good idea because bigger hierarchies of inheritance are a code smell. 
I understand your objection with the general purpose ... but if i see it correctly then the MainPanel is currently the only usecase for the AbstractContentProvider. Rename it to something like MainPanelContentProvider would fix this problem. And i think it is ok to see position and player updates as a general feature of MainPanelContentProvider implementations.
As an alternative it would also be possible to provide interfaces that can be implemented by AbstractContentProvider implementations on demand and then be checked if they are implemented by the MainPanel.
Another alternative would be a update method with a event like parameter. Then different events (PositionUpdatedEvent or PlayerUpdatedEvent) could be passed.

File: jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java (hunk 9 above)
Comment (hetzge) [reply jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java:368]: The more i am thinking about this i see the same problem here like with the Ui updaters ;) Every partition is managing its own timer instead of one timer that updates all partitions.

I agree with you that the MovableManager.getAllMovables() here feels a little bit wrong. But i am not sure how to separate this loop in a good way. The rest of the logic of updateProfessionSettings() should be moved to the ProfessionSettings class because it mainly operates on it. If the loop would be outside of PartitionManager then we would need methods that consume single moveables (like count(Moveable)). Another way would be to group the moveables by the partition and then pass all moveables of one partition at once. But i don't know if this is a good idea in the terms of performance.

Comment (paulwedeck) [reply jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java:368]: The main issue is that nobody really knows if and how long a movable stays on a partition. They could be converted, killed, pushed away (the biggest problem) or created at any time. We could hook different functions to see if their partition changes but this seams to be a too deep integration between partition and movable logic to me.


I thought about a new timerable in the PartitionsGrid that basically does:
for(mov in all movables) {

if(!mov.getMovableType().isPlayerControllable()) continue;

 partition = getPartitionAt(mov.getPosition);

if(mov.getPlayer().getPlayerId() != partition.playerId) continue;

settings = partition.getProfessionSettings();

switch(mov.getMovableType()) {
case BEARER:
settings.incBearerCount();
...
}

This needs just one iteration per movable and is not dependent on the number of partitions (in terms of complexity).

Comment (hetzge) [reply jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java:368]: > The main issue is that nobody really knows if and how long a movable stays on a partition. They could be converted, killed, pushed away (the biggest problem) or created at any time. We could hook different functions to see if their partition changes but this seams to be a too deep integration between partition and movable logic to me.
Totally agree. It would be very hard to not miss some case.

> I thought about a new timerable in the PartitionsGrid that basically does:
Yes and then remove the timer from each partition.

I will try to implement this.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 1 above)
Comment (paulwedeck) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:71]: regarding the ACP refactoring:
I might be wrong but all classes that request the position actually just want the partition, don't they?
In that case it would be an onUpdatedPartition() function.
The thing that is most important to me is that there is just one place in the code where the partition is checked against the current playerId. This should prevent other players from changing or seeing your settings and statistics without any potential loopholes or duplicate code that needs bug fixes in different places.

AbstractContentProvider -> PartitionDependentContentProvider -> WhateverPanel

This hierarchy isn't very deep for my taste and every class has its own distinct purpose.

but about the main topic:
I think my problem with your timer+absolute value is that
the arrows have their use case in increasing or decreasing some value
but only the value holder (in this case the partition) itself decides what the current value actually is and therefore what plus or minus five percent of the current value is as an absolute value.
The ProfessionsPanel does not actually know what the value will be in the future as its not related to the partition.

If you press the up arrow one or two times the ProfessionsPanel will show `x x+5 (x+10)`
but the partition might not apply the changes in five seconds because the game is paused or just very unresponsive.
In this case the ProfessionsPanel will just show the old value `x` again.
If you now press the up arrow and unpause the game afterwards the ProfessionsPanel will show and apply `x+5`.
This wouldn't make sense for my as a player if I press the same button three times but it changes just as if I pressed only one time.

Comment (hetzge) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:71]: > AbstractContentProvider -> PartitionDependentContentProvider -> WhateverPanel

The problem with this hierarchy is that you add a hierarchy step to add a feature. If you would add more features this way then your hierarchy would end up bigger and bigger. It is not the right answer to add features. A better tool to add features is composition or subtyping (not to be confused with implementation inheritance). You can read more about this here:
- https://www.yegor256.com/2016/09/13/inheritance-is-procedural.html
- https://www.infoworld.com/article/2073649/why-extends-is-evil.html

> If you press the up arrow one or two times the ProfessionsPanel will show x x+5 (x+10)
but the partition might not apply the changes in five seconds because the game is paused or just very unresponsive.
In this case the ProfessionsPanel will just show the old value x again.
If you now press the up arrow and unpause the game afterwards the ProfessionsPanel will show and apply x+5.
This wouldn't make sense for my as a player if I press the same button three times but it changes just as if I pressed only one time.

In my solution the value is directly updated in the uis data model (without any delay) and then later synced with the value of the real model. Only in very rare cases there should be a difference between the uis data model and the value from the real model. In this case the values will be back in sync after a short delay. I think this is acceptable and better then when the user has no feedback or a very late feedback caused by a paused or unresponsive game.

File: jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java (hunk 9 above)
Comment (paulwedeck) [reply jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java:368]: There is no real need to merge the execution of partitions:
- All tasks are executed in sequence (no synchronization between movable updater and partition update needed)
- its not any significant overhead for the timer
- its probably smaller code than iterating over every partitionObject
- it makes more sense for every partition to manage its own business

Comment (hetzge) [reply jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java:368]: You are right there are some things that could be a problem when looping centrally to update all partitions. So i will implement only the moveable loop in the PartitionGrid.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 1 above)
Comment (paulwedeck) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:71]: My problem is that the ui is not supposed to have a data model. The ui is only about showing and editing settings.
Your system does two things: show the current value or show what value is going to get applied next without any difference between these two indicators.
You system also does not update the value when the camera changes partition after you clicked the arrow buttons.
Your system has only one actual benefit and that would be a little bit more responsiveness if you change the bearer limit, etc. and then do something other than change it on another partition.
This is probably not an issue in most cases but will frequently happen for some players on some maps (with two islands per player).

Yes, the current way its a bit unresponsive and might lead to irritation at first but its overall cleaner in design and doesn't have any buggy edge cases (as far as I am aware).
Also reducing the UPDATE_INTERVAL in UiContentUpdater (as long as it still exists) helps a lot with the responsiveness.

File: jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java (hunk 9 above)
Comment (hetzge) [reply jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java]: It looks like this is not working. If i extract the logic to the grid then i face the following problems:
- Who is managing (starting/stopping) the timer ? Starting it in the constructor is not a good idea.
- It is necessary to iterate all partitions to reset them first (not a real problem but makes it a little bit less good)

So i will keep it like it is and just cleaned up the logic a little bit more. If you want you can also try but i think we can let it like it is for now.

Comment (paulwedeck) [reply jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java]: I was talking about an IScheduledTimerable as part of the global RescheduleTimer.
The RescheduleTimer gets saved and restored if you save/load a match and automatically started/setup or stopped when you open or close a match.

> It is necessary to iterate all partitions to reset them first (not a real problem but makes it a little bit less good)
I haven't thought about that but that shouldn't be an issue because there is the partitionObjects array in PartitionsGrid.

Comment (hetzge) [reply jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java]: Ah ok is see what my problem was. I was to much into the ui update logic. 

The RescheduleTimer.add(this, 1); should not be in the constructor. Adding something in constructor to global state is not a good idea. Instead the timer should be passed as paramter or the add should be externally.

In the loop the contains method of the PartitionManager could be reused.

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java (hunk 1 above)
Comment (hetzge) [reply jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java]: I removed the direct ui update hack and reduced the UPDATE_INTERVAL to 100. 
You are right it is better to have a unified design here.

File: jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java (hunk 9 above)
Comment (paulwedeck) [reply jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java]: I agree that you shouldn't edit global state in the constructor but its not really the case here because the PartitionsGrid is only created in the constructor of MainGrid that one is only called to setup a new match. These constructors are just a better setup() function.

I think contains() would do mostly duplicate checks because we already know that the movable is not controllable and its part of the partition (and it would cost 3-4 more function calls per movable).

File: jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java [hunk 10]
```diff
@@ -155,18 +151,14 @@ public SettlerPanel(float width, EProfessionType type) {
 
 				add(Panel.box(new CountArrows(() -> {
 					if (getTotalRatio() <= 1.0f - 0.05f) {
-						setRatio(Math.min(1f, ratio + 0.05f));
-						lastChangeTimestamp = System.currentTimeMillis();
-						SettlerPanel.this.update();
+						this.ratio = Math.min(1f, ratio + 0.05f);
 						return new SetMoveableRatioAction(type.moveableType, position, ratio);
```
Comment (paulwedeck) [inline jsettlers.graphics/src/main/java/jsettlers/graphics/map/controls/original/panel/content/settlers/profession/ProfessionPanel.java:155]: This still has the issue that the new value based of the saved value.
If you double click on one arrow you can see in the console that two events are send but they both contain the same value.
This is why I said that you shouldn't send the ratio but the delta.
for example:
new ChangeMovableRatioAction(EActionType.INCREASE_MOVABLE_RATIO, type.movableType, position)
